### PR TITLE
Add integration test for reconcile interval and update interval config

### PR DESCRIPTION
**Pull request by yhaliaw**
* Fix the broken reconcile-interval and update-interval for juju 3.x.
* Add integration test for reconcile-interval and update-interval.
* Run `apt-get clean` upon runner setup to release some disk space.
*  Add juju 2.9 integration tests. This double the number of integration tests, but we need to support juju 2.9 and 3.x.
* The reconcile-interval and update-interval charm configuration are now float type instead of int type.

### CODE REVIEW

File: .github/workflows/integration_tests.yaml [hunk 1]
```diff
@@ -64,17 +69,135 @@ jobs:
...
+      - name: Run Integration tests
+        run: |
+          tox -e integration -- \
+            --keep-models \
+            --path ${{secrets.E2E_TESTING_REPO}} \
+            --token ${{secrets.E2E_TESTING_TOKEN}} \
+            --token-alt ${{secrets.E2E_TESTING_TOKEN_ALT}} \
+            tests/integration/test_charm_scheduled_events.py
+  integration-test-charm-no-runner-juju-2-9:
+    name: Integration test charm with no runner with juju 2.9
```
Comment (jdkandersson) [inline .github/workflows/integration_tests.yaml:93]: Would it be possible to use matrix for versions of juju we want to test against? E.g., at some point we might need to test 3.2, 3.1 and 2.9 etc

File: tests/integration/conftest.py [hunk 2]
```diff
@@ -190,3 +193,65 @@ async def app(model: Model, app_no_runner: Application) -> AsyncIterator[Applica
...
+    # Spawn one runner. Spawning one runner during deployment will cause timeout waiting for model.
+    unit = application.units[0]
+
+    await application.set_config({"virtual-machines": "1"})
+    action = await unit.run_action("reconcile-runners")
+    await action.wait()
+    await model.wait_for_idle(status=ACTIVE_STATUS_NAME)
+
+    # Wait until there is one runner.
+    await assert_num_of_runners(unit, 1)
```
Comment (jdkandersson) [inline tests/integration/conftest.py:255]: `assert_...` reads like a check compared to a wait function, should this be renamed?

File: tox.ini [hunk 3]
```diff
@@ -110,3 +110,16 @@ deps =
     -r{toxinidir}/requirements.txt
 commands =
     pytest -v --tb native --ignore={[vars]tst_path}unit --log-cli-level=INFO -s {posargs}
+
+[testenv:integration-juju-2-9]
```
Comment (jdkandersson) [inline tox.ini:114]: I think tox supports parametrizing the version of a package, see: https://github.com/jdkandersson/flake8-docstrings-complete/blob/d63a6d833912f773ed2517fa4cf675686b7b91e8/tox.ini#L53

File: tests/integration/conftest.py (hunk 2 above)
Comment (yhaliaw) [reply tests/integration/conftest.py]: Renamed to `wait_till_num_of_runners`

File: .github/workflows/integration_tests.yaml (hunk 1 above)
Comment (yhaliaw) [reply .github/workflows/integration_tests.yaml]: Done.

File: tox.ini (hunk 3 above)
Comment (yhaliaw) [reply tox.ini]: Done

File: tox.ini [hunk 4]
```diff
@@ -98,24 +98,19 @@ deps =
 commands =
     bandit -c {toxinidir}/pyproject.toml -r {[vars]src_path}
 
-[testenv:integration]
-description = Run integration tests
+
+[testenv:test-flake8{5,6}]
```
[the remark this answers is not in the document]
Comment (yhaliaw) [reply tox.ini]: Removed

**Review by jdkandersson — DISMISSED** [review]

File: config.yaml [hunk 5]
```diff
@@ -42,17 +42,20 @@ options:
...
   reconcile-interval:
-    type: int
-    default: 10
+    type: float
+    default: 10.0
     description: >
       Minutes between each reconciliation between the current state and the target state of the
       runners. On reconciliation the charm polls the state of runners, such as, the number of
-      container runners, and see if actions are needed.
+      container runners, and see if actions are needed. A jitter of 25% is added. Changes to this
```
Comment (yanksyoon) [inline config.yaml:50]:
```suggestion
container runners, and see if actions are needed. A random delay of 25% is added. Changes to this
```
Minor feedback, I've had this feedback on the word `jitter` before that it's a network related term :D

File: config.yaml [hunk 6]
```diff
@@ -42,17 +42,20 @@ options:
...
+      time interval will apply when next interval is triggered.
   update-interval:
-    type: int
-    default: 60
+    type: float
+    default: 60.0
     description: >
-      Minutes between each check for new versions of the runner binary.
+      Minutes between each check for new versions of the runner binary and services the charm
+      depends on. A jitter of 25% is added. Changes to the time interval will apply when next
```
Comment (yanksyoon) [inline config.yaml:57]:
```suggestion
depends on. A random delay of 25% is added. Changes to the time interval will apply when next
```
same suggestion as above :)

File: src/event_timer.py [hunk 7]
```diff
@@ -35,6 +35,8 @@ class EventTimer:
         unit_name (str): Name of the juju unit to emit events to.
     """
 
+    _juju_run_path = Path("/usr/bin/juju-run")
+    _juju_exec_path = Path("/usr/bin/juju-exec")
```
Comment (yanksyoon) [inline src/event_timer.py:39]: May I ask where these are used? I see the additions but I don't see any references to the internal variables (perhaps github rendering issue?)

File: tests/integration/conftest.py [hunk 8]
```diff
@@ -187,6 +190,68 @@ async def app(model: Model, app_no_runner: Application) -> AsyncIterator[Applica
...
+    token: str,
+    http_proxy: str,
+    https_proxy: str,
+    no_proxy: str,
+) -> AsyncIterator[Application]:
+    """Application with no token.
+
+    Test should ensure it returns with the application having one runner.
+
+    This fixture has to deploy an new application. The scheduled events are set
```
Comment (yanksyoon) [inline tests/integration/conftest.py:213]: Minor typo suggestion :D
```suggestion
    This fixture has to deploy a new application. The scheduled events are set
```

File: tests/integration/test_charm_scheduled_events.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+
+@pytest.mark.asyncio
+@pytest.mark.abort_on_fail
+async def test_update_interval(model: Model, app_scheduled_events: Application) -> None:
+    """
+    arrange: An working application with one runner.
+    act:
+        1. Remove runner binary.
+        2. Wait for 3 minutes.
```
Comment (yanksyoon) [inline tests/integration/test_charm_scheduled_events.py:34]: Out of curiosity, may I ask where 3 minutes comes from?

**Review by yanksyoon — DISMISSED** [review]
Minor nitpicky comments and questions, LGTM!

**Review by jdkandersson — APPROVED** [review]

File: src/event_timer.py (hunk 7 above)
Comment (yhaliaw) [reply src/event_timer.py]: Removed

File: tests/integration/test_charm_scheduled_events.py (hunk 9 above)
Comment (yhaliaw) [reply tests/integration/test_charm_scheduled_events.py:34]: The intervals for scheduled events are 2 mins with a bit of random delay. So 3 mins was a safe choice.

**Review by yanksyoon — APPROVED** [review]

**Pull request merged**